[2025-07] Add order filtering support to Customer Account API /account/orders route

juanpprieto opened the pull request:
### WHY are these changes introduced?

Fixes #3079

<img width="1332" height="1506" alt="Screenshot 2025-08-27 at 8 50 24 AM" src="https://github.com/user-attachments/assets/39f16d10-b0bc-40d8-8505-27445c06f21e" />
<img width="1068" height="1496" alt="Screenshot 2025-08-27 at 8 50 16 AM" src="https://github.com/user-attachments/assets/1c1e22d4-8232-47b2-b90e-4c7c1bdd0c31" />
<img width="1330" height="1256" alt="Screenshot 2025-08-27 at 9 28 40 AM" src="https://github.com/user-attachments/assets/ab0b1c78-020a-422f-8b2b-649205d0e97d" />

The Customer Account API 2025-07 release added support for filtering orders using a `query` parameter. This allows merchants and customers to search through their order history more efficiently. Previously, customers had to manually scroll through paginated results to find specific orders.

### WHAT is this pull request doing?

This PR implements order filtering functionality in the skeleton template's customer account orders page:

- Added `query` parameter support to the `CustomerOrdersQuery` GraphQL query
- Created utility functions for building and parsing order filter parameters:
  - `buildOrderSearchQuery`: Constructs the query string for the GraphQL parameter
  - `parseOrderFilters`: Parses URL search parameters into filter objects
- Added a search form UI with inputs for:
  - Order number (automatically strips # symbol if entered)
  - Confirmation number
- Implemented URL parameter persistence for filters
- Added appropriate empty states for filtered vs unfiltered scenarios
- Added proper TypeScript types for the loader data

### HOW to test your changes?

1. Start the dev server with Customer Account Push enabled:
   ```bash
   cd templates/skeleton
   SHOPIFY_HYDROGEN_FLAG_CUSTOMER_ACCOUNT_PUSH=true npm run dev
   ```

2. Use a `.tryhydrogen.dev` URL for OAuth to work properly

3. Navigate to the customer account orders page (`/account/orders`)

4. Test filtering:
   - Enter an order number (e.g., "1021" or "#1021")
   - Enter a confirmation number (Don't know where to get these from email?)
   - Try both filters together
   - Click "Clear Filters" to reset

5. Verify:
   - Filtered results display correctly
   - URL parameters persist on navigation
   - Empty state shows when no orders match
   - Clear filters returns to full order list

⚠️ **Note**: This feature requires Customer Account API 2025-07. OAuth authentication is required for testing.

#### Post-merge steps

None required.

#### Checklist

- [x] I've read the [Contributing Guidelines](https://github.com/Shopify/hydrogen/blob/main/CONTRIBUTING.md)
- [x] I've considered possible cross-platform impacts (Mac, Linux, Windows)
- [x] I've added a [changeset](https://github.com/Shopify/hydrogen/blob/main/CONTRIBUTING.md#changesets) if this PR contains user-facing or noteworthy changes
- [ ] I've added [tests](https://github.com/Shopify/hydrogen/blob/main/CONTRIBUTING.md#testing) to cover my changes
- [ ] I've added or updated the documentation

### Additional Notes

- **Styling**: The search form currently uses inline styles for basic layout. This should be updated to match the template's design system in a follow-up.
- **Testing needed**: This implementation should be tested with subscription orders and orders with discounts to ensure filtering works correctly in all scenarios.
- **Future improvements**: Consider adding debouncing to the search inputs to reduce unnecessary re-renders.

juanpprieto committed:
[2025-07] Update Storefront API and Customer Account API

Updated both Storefront API and Customer Account API from version 2025-04 to 2025-07

- Updated version constants in hydrogen-react and hydrogen packages
- Regenerated GraphQL types and schemas for both APIs
- Updated all hardcoded version references in documentation and tests
- Regenerated skeleton template types
- Created 6 GitHub issues for tracking API changes implementation

- #3076: Subscription discount data in Customer Account API
- #3077: Cart warnings for non-applicable discount codes
- #3078: BUYER_CANNOT_PURCHASE_FOR_COMPANY_LOCATION cart error
- #3079: New order filter options in Customer Account API
- #3080: Imperial units in UnitPriceMeasurement
- #3081: Selling plan error handling in Cart mutations

- TypeScript: ✅ No errors
- Tests: ✅ All passing
- Lint: ⚠️ 2 warnings (non-blocking import duplicates)
- Build: ✅ Successful

juanpprieto committed:
Update package-lock.json for @shopify/cli dependency update

juanpprieto committed:
[2025-07] Fix Monorepo RR7 CI build failures and improve reliability (#3091)

* fix: Complete React Router 7 migration for Express example

The Express example was partially migrated to React Router 7 but missing critical
configuration and dependencies, causing CI failures.

Changes:
- Add react-router.config.ts with proper preset and app directory configuration
- Fix vite.config.ts to call reactRouter() without arguments (RR7 API change)
- Add @react-router/dev as devDependency for TypeScript types
- Update entry.server.tsx to remove unnecessary @remix-run/web-fetch import
  (Node 20+ provides native Response globally)
- Fix environment variable loading using getEnv() helper
- Add .gitignore for React Router generated type files
- Update tsconfig.json to exclude generated directories

The example now builds successfully with React Router 7's framework mode.

* fix: Fix CLI build checks and TypeScript configuration

Multiple CLI-specific issues were causing CI failures after recent refactoring.

Changes:
- Fix build-check.mjs to expect .jsx extensions (tsup outputs .tsx→.jsx)
- Exclude assets/routes from TypeScript checking (template files for user projects)
- Align ast-grep to version 0.33.0 to match @shopify/cli's bundled version
- Fix isHydrogenMonorepo() detection to work in both dev and npm package contexts
- Add 'as const' assertions for ast-grep 0.33.0 API compatibility

Background:
- Template files were moved to assets/routes in May 2024 and inadvertently
  included in TypeScript scope
- ast-grep 0.34.1 conflicted with @shopify/cli's bundled 0.33.0

* fix: Fix CI workflow and monorepo build orchestration

The CI workflow and Turbo configuration had multiple issues causing intermittent
build failures, particularly the "No lockfile found" error.

Changes:
- Add SHOPIFY_HYDROGEN_FLAG_LOCKFILE_CHECK=false to typecheck step in CI
  (was only on build:all, but typecheck also triggers builds via Turbo)
- Add SHOPIFY_HYDROGEN_FLAG_LOCKFILE_CHECK to turbo.json globalEnv
- Add explicit build dependencies for all examples in turbo.json
- Remove docs-preview from workspaces to prevent React Router version conflicts
- Regenerate package-lock.json for dependency resolution

The root cause was that the typecheck step depends on build tasks (per turbo.json)
but wasn't receiving the lockfile check flag, causing skeleton builds to fail.
This has been a source of CI instability since the monorepo migration.

juanpprieto committed:
[2025-07] Fix Money component compatibility with Customer Account API USDC currency (#3090)

* Fix Money component compatibility with Customer Account API USDC currency

The 2025-07 API update introduced USDC currency code to Customer Account API but not Storefront API, causing TypeScript errors and runtime failures.

Changes:
- Update Money component to accept MoneyV2 from both Storefront and Customer Account APIs via union types
- Enhance useMoney hook to detect unsupported currency codes (like USDC) and gracefully fall back to decimal formatting
- Add currency code suffix for unsupported currencies (e.g., "100.00 USDC") to maintain clarity
- Default to 2 decimal places for USDC to reinforce its 1:1 USD peg based on industry standards

Technical details:
- Handle Intl.NumberFormat RangeError for cryptocurrency codes not in ISO 4217
- Add comprehensive test coverage for both API types and USDC formatting
- Update TypeScript types to support both CurrencyCode enums

Fixes #3089

* Add changeset for Money component USDC compatibility fix

* Add build script to copy customer-account-api-types to dist

Since our Money component now imports from customer-account-api-types, we need to ensure this file is copied to the dist folder during the build process. Previously, only storefront-api-types was being copied.

- Replace copy-storefront-types with copy-api-types script that copies both API type files
- Update build and dev:demo scripts to use the new copy-api-types script
- This fixes the CI build failure where customer-account-api-types.d.ts was not found in dist

* Trigger CI rebuild

kdaviduik committed:
Suggested money changes for PR #3090 (#3097)

juanpprieto committed:
[2025-07] Fix React Context error during client-side hydration (#3101)

* Fix React Context errors in Vite dev server

This fixes "Cannot read properties of null (reading 'useContext')" errors that occur when React Context hooks are used in development mode.

Root Cause:
Vite's module federation can cause multiple React instances to be loaded when @shopify/hydrogen is served as an external module. React Context requires a single React instance to function properly - when multiple instances exist, Context providers from one instance cannot be consumed by hooks from another instance, resulting in null context values.

Solution:
1. Added resolve.dedupe for react, react-dom, and @shopify/hydrogen to ensure Vite uses a single instance of these packages
2. Modified optimizeDeps.include to conditionally optimize @shopify/hydrogen:
   - In production/regular projects: Optimize @shopify/hydrogen to prevent context errors
   - In monorepo development: Skip optimization to allow live reload of source changes

This maintains development workflow in the monorepo while fixing the production build issue.

* Fix React Context error during client-side hydration

Resolves 'Cannot read properties of null (reading useContext)' error
that occurred on first page load when using CSP with nonces.

The issue was caused by NonceProvider being present during SSR but
missing during client hydration, creating a React Context mismatch.

Changes:
- Export NonceProvider from @shopify/hydrogen for client-side usage
- Add NonceProvider wrapper to skeleton's entry.client.tsx
- Simplify Vite config to improve React Context stability

juanpprieto committed:
Add order filtering support for Customer Account API 2025-07

Implement new order filtering capabilities using the query parameter
introduced in Customer Account API version 2025-07. Customers can now
filter orders by name (order number) and confirmation number.

Changes:
- Add query parameter to CustomerOrdersQuery GraphQL fragments
- Create orderFilters utility functions for building search queries
- Add search form UI to orders page with inputs for filters
- Update empty state to handle filtered vs unfiltered scenarios
- Persist filter state in URL parameters for bookmarking

The implementation follows the new filtering syntax:
- name:1001 for order number filtering
- confirmation_number:ABC123 for confirmation number filtering
- Combined filters using AND operator

kdaviduik committed:
Update Vitest from v1.0.4 to v3.2.4 across all packages (#3126)

* Update Vitest from v1.0.4 to v3.2.4 across all packages

- Upgrade vitest to ^3.2.4 in all 5 packages that use it
- Upgrade @vitest/coverage-v8 to ^3.2.4 to match vitest version
- Add vitest as devDependency to create-hydrogen package (was missing)
- Add explicit coverage provider configuration (required in v3+)
  - Set provider: 'v8' in cli and hydrogen-react vitest configs

Breaking change: Vitest v2+ requires explicit coverage provider config

* Fix Vitest v3 Mock type breaking changes

In Vitest v3, vi.fn() returns Mock<() => unknown> instead of Mock<any[], any>,
requiring explicit type assertions or type parameters for mocks with specific
return types.

Changes:
- Add type assertion for fetchWithServerCache mock in storefront.test.ts
- Simplify Promise creation in session mocks (Promise.resolve instead of new Promise)
- Fix Mock type usage in Image.test.tsx for console.warn assertions
- Add vitest/globals to tsconfig for packages using global test functions
- Keep jest types alongside vitest types for @testing-library/jest-dom compatibility

The stricter typing in v3 catches more potential issues at compile time but
requires updating existing mocks that relied on v1's permissive typing.

* Fix formatting

juanpprieto committed:
Improve order filtering UI with minimal, mobile-friendly design

- Simplified HTML structure using semantic fieldset/legend elements
- Removed inline styles in favor of CSS classes
- Added responsive grid layout (single column mobile, two column desktop)
- Improved accessibility with proper ARIA labels and search input types
- Reduced visual complexity while maintaining full functionality
- Minimal CSS that allows easy customization by developers

juanpprieto committed:
Add changeset

juanpprieto committed:
Address PR feedback for order filtering feature

- Fix duplicate empty state logic by passing filters to OrdersTable
- Add input sanitization to prevent injection attacks
- Extract field name constants for maintainability
- Remove hardcoded string repetition throughout component

kdaviduik commented on templates/skeleton/app/lib/orderFilters.ts:
```diff
@@ -0,0 +1,84 @@
+/**
+ * Field name constants for order filtering
+ */
+export const ORDER_FILTER_FIELDS = {
```
FYI there are WAY more filter fields than just these 2 https://shopify.dev/docs/api/customer/latest/queries/customer#returns-Customer.fields.orders.arguments.query. We don't need to add support for all filters in this PR though (or for this release).

rbshop replied:
For now I'm going to add a link to the full set of filters in a comment. I'm not sure we should be implementing an exhaustive list here as part of the skeleton, but if we do decide to do that then we can in a follow up PR

rbshop replied:
Done in https://github.com/Shopify/hydrogen/pull/3125/commits/37df1c4b5d76dde7c3c6ad501b0a1fa70d83f128

kdaviduik commented on templates/skeleton/app/lib/orderFilters.ts:
```diff
@@ -0,0 +1,84 @@
+ * @param value - The input string to sanitize
+ * @returns The sanitized string
+ */
+function sanitizeFilterValue(value: string): string {
```
Imo we should definitely have unit tests for all of this stuff as part of this PR. 

In terms of the end state, this is something that's a pain to manually test and these will very objectively be "working" or "not working" with the API (which simpler tests can't guarantee), so I think we absolutely should add E2E tests for stuff like this. Doesn't need to happen as part of this PR/release though

rbshop replied:
I agree that we want E2E tests for the skeleton as a whole, and especially for this. As for unit tests though, historically we have made the explicit decision to not include unit tests in the skeleton (IIRC the reasoning was to optimise initial developer velocity).

juanpprieto replied:
💯 we will address this and all recipes with e2e we just need to setup an infrastructure than can support it SFAPI / CAAPI auth / orders / subs.. etc

kdaviduik commented on templates/skeleton/app/lib/orderFilters.ts:
```diff
@@ -0,0 +1,84 @@
+function sanitizeFilterValue(value: string): string {
+  // Only allow alphanumeric, underscore, and dash
+  // Remove anything else to prevent injection
+  return value.replace(/[^a-zA-Z0-9_\-]/g, '');
```
This specific regex for sanitation seems 👍🏻 

<details>

<summary>Claude's verification</summary>

## Verification Results: Current Sanitization is SAFE ✅

Based on my comprehensive research and testing:

### **Valid Characters in Shopify Fields:**

1. **Order Names**: `#1001`, `Store1001` 
   - Uses: letters, numbers, hash (which is stripped separately)

2. **Confirmation Numbers**: `XPAV284CT`, `R50KELTJP`, `35PKUN0UJ`
   - Uses: uppercase letters and numbers only

3. **Order Numbers**: Pure integers like `1001`

### **Current Sanitization (`/[^a-zA-Z0-9_\-]/g`):**

✅ **Preserves all valid Shopify formats**
- Keeps alphanumeric (covers all real examples)
- Keeps underscore and dash (for extended formats like `ORD-2024-001`)
- Hash (`#`) is handled separately by stripping before sanitization

✅ **Blocks ALL injection attempts**
- Removes quotes, backslashes, brackets
- Removes SQL/GraphQL operators (`:`, `>`, `<`, `=`)
- Removes template syntax (`${}`, `{{}}`)
- Removes all XSS vectors

### **Key Insight:**

The sanitization appears "too restrictive" because it removes colons, BUT that's actually correct! The colons in `name:1001` are part of the **query syntax**, not the user input. The code correctly:

1. Sanitizes user input: `"1001"` → `"1001"`
2. Then builds query: `"name:" + "1001"` → `"name:1001"`

### **Conclusion:**

The current implementation is **production-ready and secure**. It:
- ✅ Works with all documented Shopify formats
- ✅ Prevents GraphQL injection effectively
- ✅ Follows defense-in-depth principles

The only minor limitation is removing spaces from order names, but Shopify's actual order formats don't use spaces in the examples, so this is acceptable.


</details>

kdaviduik commented on templates/skeleton/app/routes/account.orders._index.tsx:
```diff
@@ -35,45 +51,131 @@ export async function loader({request, context}: LoaderFunctionArgs) {
+          Filter Orders
+          {hasFilters && (
+            <small className="order-search-active">
+              ({[currentFilters.name, currentFilters.confirmationNumber].filter(Boolean).length} active)
```
This `"<number> active"` UX is really confusing imo - even if I delete the content from both filters, it doesn't update. I would personally just delete this

<img width="474" height="320" alt="Image" src="https://github.com/user-attachments/assets/bb92adef-d808-4a50-b73a-0e3e7654e6dd" />

rbshop replied:
Done in https://github.com/Shopify/hydrogen/pull/3125/commits/84718a1e9be686d0c9185650870f5088ec65ee43

kdaviduik commented on templates/skeleton/app/routes/account.orders._index.tsx:
```diff
@@ -35,45 +51,131 @@ export async function loader({request, context}: LoaderFunctionArgs) {
 }
 
-function OrdersTable({orders}: Pick<CustomerOrdersFragment, 'orders'>) {
+function OrdersTable({orders, filters}: {orders: CustomerOrdersFragment['orders']; filters: OrderFilterParams}) {
```
We should definitely have an `aria-live: polite` on this for accessibility

rbshop replied:
I saw that this was already addressed in your commit (based on the commit message), but I think that was maybe lost during a rebase, so it has been added in https://github.com/Shopify/hydrogen/pull/3125/commits/37df1c4b5d76dde7c3c6ad501b0a1fa70d83f128

kdaviduik commented on templates/skeleton/app/routes/account.orders._index.tsx:
```diff
@@ -35,45 +51,131 @@ export async function loader({request, context}: LoaderFunctionArgs) {
+        <div className="order-search-buttons">
+          <button type="submit">Search</button>
+          {hasFilters && (
+            <button 
```
I think this button should also clear the field inputs (in addition to or instead of clearing the URL query params), otherwise the UX is confusing

<img width="511" height="417" alt="Image" src="https://github.com/user-attachments/assets/ab58a1c7-0ff9-4876-9b60-80a24a4a27e3" />

rbshop replied:
Done in https://github.com/Shopify/hydrogen/pull/3125/commits/84718a1e9be686d0c9185650870f5088ec65ee43

kdaviduik commented on templates/skeleton/app/routes/account.orders._index.tsx:
```diff
@@ -35,45 +51,131 @@ export async function loader({request, context}: LoaderFunctionArgs) {
+        </div>
+        
+        <div className="order-search-buttons">
+          <button type="submit">Search</button>
```
I don't want to be too picky about the UX but I think very basic "loading" state functionality should be here as part of this PR. I like the loading UX for the "Profile" form where it simply disables the button + adds "ing" to it while loading. I think we should copy that here

rbshop replied:
Done in https://github.com/Shopify/hydrogen/pull/3125/commits/84718a1e9be686d0c9185650870f5088ec65ee43

kdaviduik commented on templates/skeleton/app/lib/orderFilters.ts:
```diff
@@ -0,0 +1,84 @@
+  /** Order name or number (e.g., "#1001" or "1001") */
+  name?: string;
+  /** Order confirmation number */
+  confirmationNumber?: string;
```
[We don't seem to be querying or, therefore, displaying the order confirmation number](https://github.com/Shopify/hydrogen/blob/ed0c8b5169cb6e45a4ebe970cd1e80e07c8f7769/templates/skeleton/app/graphql/customer-account/CustomerOrdersQuery.ts#L4-L17). As a result, it's confusing that we are offering the option to filter by this. 

Personally I think we should delete this option and just have `name` as a filter param for now. I'd also be fine with replacing this with a different filter param that we're rendering (or start querying + rendering the order confirmation number and keep this filter param), but deleting this is the simplest option for now. Up to you.

rbshop replied:
Confirmation number is now being queried as of https://github.com/Shopify/hydrogen/pull/3125/commits/84718a1e9be686d0c9185650870f5088ec65ee43

kdaviduik committed:
PR feedback

- Add confirmationNumber to orders query
- Render confirmationNumber for orders in list, and on order details page
- Clear both form inputs when the "Clear" button is pressed
- Delete the "<number> active" that stated how many filters were active but was a bit confusing
- Added aria-live=polite to orders table for accessibility
- Disable buttons and change text from "Search" to "Searching" when form has been submitted
    - As part of this, I also removed the styles on these buttons. The default browser styles are nicer here (ie: it greys out disabled buttons)

rbshop committed:
Undoing a couple of accidental non-changes from rebasing

rbshop committed:
Link to full list of available fields, add aria-live polite

juanpprieto committed:
Update subscriptions recide

juanpprieto committed:
update lock

juanpprieto committed:
Update lock

juanpprieto committed:
Fix recipe base commit reference to point to commit where recipe was created

juanpprieto committed:
Update recipe to use commit from main where it was merged

juanpprieto committed:
Fix legacy-customer-account-flow recipe for React Router 7.8.x compatibility

- Update all patches to use React Router 7.8.x consolidated imports
- Complete conversion from Customer Account API to Storefront API
- Fix account.orders.$id.tsx and account.orders._index.tsx patches
- Add missing account.profile.tsx route for legacy customer flow
- Update recipe.yaml with regenerated patch hashes
- Ensure all routes use session-based customer access tokens

juanpprieto committed:
feat(multipass): align multipass recipe with legacy customer account flow

- Apply all order-related patches from legacy recipe to multipass
- Fix redirect loop by using Storefront API instead of Customer Account API
- Update account.orders._index.tsx to use session-based auth with proper GraphQL query
- Update account.orders.$id.tsx with complete order details query
- Apply profile patches to support email, phone, password changes
- Ensure all routes use classic customer accounts with session tokens
- Remove conflicting Customer Account API usage in multipass context

juanpprieto committed:
revert lock

juanpprieto committed:
fix type issue

juanpprieto committed:
add cookbook deps

juanpprieto committed:
Fix legacy recipe

juanpprieto committed:
Fix cookbook recipes for React Router 7.8.x compatibility

Fixed two cookbook recipes that were failing validation:

1. legacy-customer-account-flow recipe:
   - Removed confirmationNumber references (not available in Storefront API)
   - Added proper error handling for loader data that can return errors
   - Updated patches to match current skeleton structure
   - All TypeScript errors resolved

2. partytown recipe:
   - Updated vite.config.ts patch to account for new server configuration block
   - Regenerated all patches with new hash names
   - Recipe now applies cleanly without conflicts

Both recipes now pass all validation tests including:
- npm install
- npm run codegen
- npm run typecheck
- npm run build

rbshop committed:
Fixing express recipe

rbshop committed:
Skip lockfile check during typecheck on ci

rbshop committed:
Skip lockfile check in deploy-examples workflow

rbshop commented on .github/workflows/ci.yml:
```diff
@@ -94,7 +94,7 @@ jobs:
 
       - name: ✅ Typecheck
-        run: npm run typecheck
+        run: SHOPIFY_HYDROGEN_FLAG_LOCKFILE_CHECK=false npm run typecheck
```
I added this back in because it was failing on CI due to the missing lockfile in the skeleton, so I'm very confused as to how this was passing before?

juanpprieto replied:
I have the same questions and it's not the first time I have run into this random issue as seen in some of the other similar edits in this workflow. My best guess is either workflow actions update or typescript update now being more strict and a new lock file just showing an issue that was hidden
